### ISSUE TITLE
Battery charge limits (e.g. max 85%) to improve battery health over time

### ISSUE THREAD

**Issue by sandreas** [GrapheneOS/os-issue-tracker#3213]
Usually, the battery of my phone charges until it hits 100% charging level. 

I would love to see a feature to limit the charging level to e.g. 85%. This would improve battery health over time. This setting is also present in Lenovo Notebooks.

**Comment by snrkl** [comment]
FWIW the pixel tablet restricts charging to ~80% when on the provided dock..

**Comment by sandreas** [comment]
@thestinger 
Is there any way to vote for this feature? I wonder if thumbsup clicks are recognized in any way...

**Comment by aureq** [comment]
To add a little more here... It appears there's a Magisk module ([link](https://github.com/VR-25/acc)) that already do that as well as a front-end app ([link](https://github.com/MatteCarra/AccA)). The main issue though is this solution require `root` access to the device to control the android kernel behavior.

To be confirmed, but it _feels_ like this would require a tight integration into GrapheneOS as exposing `root` is a no.

**Comment by sandreas** [comment]
@aureq 

Yeah there is[ a reddit post ](https://www.reddit.com/r/Android/comments/f4d66o/option_to_charge_to_only_80_soc_not_100/) claiming, that the indicator totally is a lie and `100%` actually represents `80%-90%` charging level to optimize the battery life and `80%` would be `64%` in reality.

However, I would love to hear that from a developer posting a code reference - I did not find something reliable anywhere.

**Comment by eugenesan** [comment]
First, that feature is an absolute and hard requirement for devices that expected to last 3+ years!
I can understand why Google won't implement it as a manual option (dumb users, AI propagation, short device upgrade cycles etc) but I can't figure why it is not a standard option on GrapheneOS especially when configuring battery charging limits on Tensor SOCs is as simple as:
```
echo 80 > /sys/devices/platform/google,charger/charge_start_level
echo 85 > /sys/devices/platform/google,charger/charge_stop_level
```
_And this is exactly how Pixel's "AI" does it to optimize the battery_

Unfortunately, that requires root access or integration in the OS.
It should pretty trivial to do and I hope there will enough pressure on Graphene to do that since that's the only thing keeping me from using GrapheneOS.

Regarding "fake" charge levels, it's incorrect.
Android's "charge control" will inform you about if and when it does the "optimization" and you'll continue to see the actual battery state of charge.

**Comment by uberhalit** [comment]
[Any news](https://discuss.grapheneos.org/d/9017-how-can-i-limit-battery-charge-to-80/37) on this? My phone gets held hostage at 100% by android auto whenever I do a road trip.

**Comment by sandreas** [comment]
> It's possible that Android 15 will include this in AOSP:

Looks like there is no slider or something, just the fixed value of 80%. Good but still a bit disappointing, because  on my (older) device  80% probably won't last the whole day. I had hoped for something custom like 85% or 90% to be able to have a value between 10% to 90%... well no reason to complain tough. Great news.

**Comment by tareefdev** [comment]
Unfortunately, it not yet there
https://9to5google.com/2024/10/16/google-pixel-battery-charging-limit-android-15/

**Comment by lukas-fichtner** [comment]
I think it is also not included in QPR1, seems to be a Google only thing: https://www.androidpolice.com/android-15-qpr1-features/

**Comment by oscarorange** [comment]
Limit to 80% was just added to stock OS with the latest update.

**Comment by oscarorange** [comment]
I see.

Does GOS aim to add support for charge limiting in some form or another at some point to remain on par feature wise to stock in that regard?

**Comment by newhinton** [comment]
Hm i have to disagree on that a bit. The general consensus is that only charging your battery ~80% truly improves battery longevity since the most stressful part of charging is omitted. (In that sense some phones even throttle charging to only reach that setpoint when an alarm was set). I have not found a primary source for that though.

But more importantly, the other thing i wanted to say is that it is likely not feasible for a lot of people to change out their batteries regularly, especially for devices that are glued together.

For those people, increasing battery-lifespan (dven while sacrificing runtime) might be an important ability to keep their devices workable longer.

**Comment by oscarorange** [comment]
I fully understand your point, but personally I'd still rather limit to ~80% knowing I'm keeping the battery capable of that 80% for as long as possible, as battery replacements are definitely something that's far too intimidating for me to try myself and are a hassle and expensive to get done professionally.

I've found ~80% is enough to get me through the day and I rarely need to charge more than once a day. Sacrificing 20% from the get go to save my time and money down the line is definitely a trade off my frugal self would make.

Man I miss the removable batteries of yesteryear.

**Comment by sandreas** [comment]
> The consequence of the battery degrading is lower battery life. Setting it to an 80% cap has already effectively degraded it to 80% from day one. 

I see you have a pretty strong opinion on this. What about the proof / math? 

There are numerous sources, unfortunately not one that I found trustworthy enough to absolutely rely on it. What I personally found out is that a battery lasts about *4 times* to *7 times* longer (depending on the device), if charged carefully (20-80%).

A battery switch for my Pixel 4a starts at 69 bucks, DIY is a PAIN to do (even with an ifixit kit), so you would save at least 138 bucks just by charging the phone between 20% to 80% over 7 years. Moreover after about 18 months the 100% battery would have the same amount of hours like the 80% charged one because the 100% one is degrading so much faster.

One of the sources: 
https://electronics.stackexchange.com/questions/623358/wouldnt-charging-a-lithium-battery-to-80-only-defeat-the-purpose-of-putting-th

**Comment by dill-shower** [comment]
> Capping charging at 80% means that 



Not all GrapheneOS users are in countries where Google provides repair services. In case of problems with the battery, you will have to look for a self-taught person who can fix it (or maybe break something in the process), so in order to definitely not break anything, the only option is to buy a new Google pixel before the end of the security update period.

Maybe it makes sense to add integration with Google services for this function for now? Since it does not look like it will be added anytime soon, it would be useful to make do with what we have for now. Better "good that works" than "perfect that may be in the unspecified distant future"

**Comment by dill-shower** [comment]
The question of whether this will improve battery performance is very difficult, but it certainly won't make it worse. At the same time, I personally almost always use the device while it is connected to the network. As a result, the battery voltage will be 4.4V all day. This is definitely not useful.There has been some kind of charging optimization for a long time, but it doesn't work in GOS. The charging speed doesn't decrease in any way, it charges the same way, 100%

**Comment by traversall** [comment]
> If you want the feature sooner, you're welcome to work on adding configuration for it. There was no blocker waiting for Android 15 QPR1. The infrastructure for it has been available for a long time and we realized the QPR1 feature is a Google app feature a while ago so it wouldn't provide anything to us.

There is an actual need for this feature.  If you don't want to provide it yourself, but you have pinpointed how Google implemented it, then please at least share enough information so that others can contribute a solution.

**Comment by eiqnepm** [comment]
> > If you want the feature sooner, you're welcome to work on adding configuration for it. There was no blocker waiting for Android 15 QPR1. The infrastructure for it has been available for a long time and we realized the QPR1 feature is a Google app feature a while ago so it wouldn't provide anything to us.
> 
> There is an actual need for this feature.  If you don't want to provide it yourself, but you have pinpointed how Google implemented it, then please at least share enough information so that others can contribute a solution.

https://github.com/GrapheneOS/os-issue-tracker/issues/3213#issuecomment-2068161800
